docs/keto/modeling/learn-the-concepts.mdx

```diff
@@ -0,0 +1,295 @@
+- Folder
+
+This list translates into our first version of the permission model.
+
+```ts title="permissions-v1.ts"
+import { Namespace, Context } from "@ory/keto-namespace-types"
+
+// highlight-start
+class User implements Namespace {}
+class Document implements Namespace {}
+class Folder implements Namespace {}
+// highlight-end
+```
+
+:::tip
+
+The objects and subjects from your description are [namespaces](../concepts/namespaces) in Ory Permissions, and are declared as
+classes in the Ory Permission Language. The convention is to spell these class names as TypeScript classes, capitalized and in
+singular form, for example `Document` instead of `documents`.
+
+:::
+
+### List relationships to other objects
+
+Read through the description sentence by sentence to find the list of relationships:
+
+> We want to model `user` access to `documents` such that `users` can be an `owner`, `editor`, or `viewer` of a `document`.
+
+<Mermaid
+  chart={`
+graph LR
+    Document -->|owner| User
+    Document -->|editor| User
+    Document -->|viewer| User
+`}
+/>
+
+> An owner is also an editor of a `document`, and an editor is also a viewer of a `document`. Furthermore, `documents` can put
+> into a hierarchy of `folders`, such that being able to view the parent `folder` implies being able to view the enclosed
+> `folders` and `documents`.
+
+<Mermaid
+  chart={`
+graph LR
+    Document -->|parent| Folder
+    Folder -->|parent| Folder
+    Folder -->|owner| User
+    Folder -->|editor| User
+    Folder -->|viewer| User
+`}
+/>
+
+Putting this all together in one graph:
+
+<Mermaid
+  chart={`
+graph LR
+    Document -->|owner| User
+    Document -->|editor| User
+    Document -->|viewer| User
+    Document -->|parent| Folder
+    Folder -->|parent| Folder
+    Folder -->|owner| User
+    Folder -->|editor| User
+    Folder -->|viewer| User
+`}
+/>
+
+### List permissions to check for each object
+
+When you perform an action on behalf of a user, you usually check a specific permission, such as `view` or `edit`, instead of a
+relationship, such as `owner`, and have the relationship implied by the permission.
+
+Add these permissions to the database to express what the application needs. For our document storage, you have these permissions:
+
+- `view` a document if the user is a `viewer`, `editor`, or `owner` of the document; or if the user can `view` the parent folder
+- `edit` a document if the user is a `editor`, or `owner` of the document; or if the user can `edit` the parent folder
+- `delete` a document if the user is an `owner` of the document; or if the user can `delete` the parent folder
+- `share` a document if the user is an `owner` of the document; or if the user can `share` the parent folder
+- `delete` a folder if the user is an `owner` of the folder; or if the user can `delete` the parent folder
+- `share` a folder if the user is an `owner` of the folder; or if the user can `share` the parent folder
+
+### Define relations in the OPL
```

Comment:
I think this should happen before the permission topic starts.

Reply:
You mean instead of

1. Create a list of objects, which are things that you want to give access to.
2. For each object
   1. Make a list of relationships this has to other objects. In a database, those would be associatons expressed with foreign keys.
   2. Make a list of permissions that you want to check.
5. Define each relation in the OPL
6. Define each permission in the OPL
7. Test your permission model

the outline would be

1. Create a list of objects, which are things that you want to give access to.
1. Make a list of relationships this has to other objects. In a database, those would be associatons expressed with foreign keys.
5. Define each relation in the OPL
3. Make a list of permissions that you want to check.
6. Define each permission in the OPL
7. Test your permission model

?

Reply:
I mean because you first determine the relationships, then you should directly add them to the OPL before you go into the permissions topic.

Reply:
agreed